src/ray/util/subreaper.cc

```diff
@@ -0,0 +1,177 @@
+#endif
+
+#include "absl/container/flat_hash_set.h"
+#include "absl/synchronization/mutex.h"
+#include "ray/util/logging.h"
+#include "ray/util/macros.h"
+#include "ray/util/process.h"
+
+namespace ray {
+// Platform-specific implementation of subreaper code.
+
+#ifdef __linux__
+// Linux implementation.
+
+namespace {
+
+using SignalHandlerFn = void (*)(const boost::system::error_code &, int);
+// Register a signal handler for the given signal.
+// The handler will be called with the signal_set and the error code.
+// After the handler is called, the signal will be re-registered.
+// The callback keeps a reference of the shared ptr to make sure it's not destroyed.
+void RegisterSignalHandlerLoop(std::shared_ptr<boost::asio::signal_set> signals,
+                               SignalHandlerFn handler) {
+  signals->async_wait(
+      [signals, handler](const boost::system::error_code &error, int signal_number) {
+        handler(error, signal_number);
+        RegisterSignalHandlerLoop(signals, handler);
+      });
+}
+
+void SigchldHandlerReapZombieAndRemoveKnownChildren(
+    const boost::system::error_code &error, int signal_number) {
+  if (error) {
+    RAY_LOG(WARNING) << "Error in SIGCHLD handler: " << error.message();
+  }
+  int status;
+  pid_t pid;
+  // Reaps any children that have exited. WNOHANG makes waitpid non-blocking and returns
+  // 0 if there's no zombie children.
+  while ((pid = waitpid(-1, &status, WNOHANG)) > 0) {
+    if (WIFEXITED(status)) {
+      RAY_LOG(INFO) << "Child process " << pid << " exited with status "
+                    << WEXITSTATUS(status);
+    } else if (WIFSIGNALED(status)) {
+      RAY_LOG(INFO) << "Child process " << pid << " exited from signal "
+                    << WTERMSIG(status);
+    }
+    KnownChildrenTracker::instance().RemoveKnownChild(pid);
+  }
+}
+
+}  // namespace
+
+// Set this process as a subreaper.
+// Only works on Linux >= 3.4.
+bool SetThisProcessAsSubreaper() {
+  if (prctl(PR_SET_CHILD_SUBREAPER, 1) == -1) {
+    RAY_LOG(WARNING) << "Failed to set this process as subreaper: " << strerror(errno);
+    return false;
+  }
+  return true;
+}
+
+void SetSigchldIgnore() { signal(SIGCHLD, SIG_IGN); }
+
+void SetupSigchldHandlerRemoveKnownChildren(boost::asio::io_context &io_service) {
+  auto sigchld_signals = std::make_shared<boost::asio::signal_set>(io_service, SIGCHLD);
+  RegisterSignalHandlerLoop(sigchld_signals,
+                            SigchldHandlerReapZombieAndRemoveKnownChildren);
+  RAY_LOG(INFO)
+      << "Raylet is set to reap zombie children and remove known children pids.";
+}
+
+// Kill all child processes that are not owned by this process.
+// It's done by checking the list of child processes and killing the ones that are not
+// created via Process::spawnvpe.
+//
+// TODO: Checking PIDs is not 100% reliable because of PID recycling. If we find issues
+// later due to this, we can use pidfd.
+void KillUnknownChildren() {
+  auto to_kill =
```

Comment:
Q: Is this automatically optimized by compiler?

Reply:
I think it can do guaranteed copy elision but not 100% sure